### DISCUSSION TITLE
Recording to WAV only results in a hissing noise

### DISCUSSION THREAD

**Discussion by Tacco-172** [pschatzmann/arduino-audio-tools#1667] [Q&A] [answered]
Hi there

I'm trying to record audio from an I2S ADC to a WAV file on an SD card for an aviation project.

Currently, I've managed to generate audio files, but the content is just hissing noise. 
The file header appears to be fine, and the file is recognized by the audio player on my computer. However, there's a repeating pattern of weirdly high values. 
Since I haven't connected anything to the ADC, I expect silence, so I'm unsure where the noise is coming from.

I suspect there's an issue with my sketch, but I haven't been able to identify the cause. For testing, I've kept the sketch as simple as possible and close to the examples, but I still can't get it to work.

I'd greatly appreciate any help or guidance in resolving this issue, as I don't have a ton of experience.

```
/**
 * Project base:
 * @file streams-i2s-serial.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-i2s-serial/README.md
 * @author Phil Schatzmann
 * @copyright GPLv3
 **/

// imports
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"
#include <SD.h>
#include <SPI.h>

// definitions
#define PIN_AUDIO_KIT_SD_CARD_CS 5
#define PIN_AUDIO_KIT_SD_CARD_MOSI 23
#define PIN_AUDIO_KIT_SD_CARD_MISO 19
#define PIN_AUDIO_KIT_SD_CARD_CLK 18
const char* file_name = "/rec.wav";
uint64_t timeout;
uint8_t rec_length = 30;  // length of record in seconds

// audio setup
AudioInfo info_i2s(48000, 2, 24);  // input stream config
AudioInfo info_sd(16000, 1, 16);   // output stream config

I2SStream i2sStream;                                   // I2S as input stream
File file;                                             // final audio output
EncodedAudioStream EncAudio(&file, new WAVEncoder());  // encode stream to wav for SD
StreamCopy copier(EncAudio, i2sStream);                // copy i2sStream to EncodedAudioStream

// arduino setup
void setup(void) {
  // serial
  Serial.begin(115200);                                         // start serial
  delay(100);                                                   // delay for serial transmission issues
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  // AudioTools serial log config

  // audio configs
  auto i2s_cfg = i2sStream.defaultConfig(RX_MODE);  // furthter i2s input config
  i2s_cfg.copyFrom(info_i2s);                       // copy config from above
  i2s_cfg.i2s_format = I2S_STD_FORMAT;              // or try with I2S_LSB_FORMAT
  i2s_cfg.is_master = true;                         // this module nees a master clock, ESP32 is master
  i2s_cfg.use_apll = true;                          // try with yes for more accurate clock
  i2s_cfg.pin_mck = 0;                              // master clock pin
  i2sStream.begin(i2s_cfg);                         // apply with config

  auto encoder_cfg = EncAudio.defaultConfig();  // further sd output config
  encoder_cfg.copyFrom(info_sd);                // copy config from above
  EncAudio.begin(encoder_cfg);                  // apply with config

  // SD setup
  if (!SD.begin(PIN_AUDIO_KIT_SD_CARD_CS)) {   // open SD drive
    Serial.println("Initialization failed!");  // stop if failed to open SD drive
    stop();                                    //
  }                                            //
  if (SD.exists(file_name)) {                  // cleanup file if necessary
    SD.remove(file_name);                      //
  }                                            //
  file = SD.open(file_name, FILE_WRITE);       // open file
  if (!file) {                                 // stop if failed to open file
    Serial.println("file failed!");            //
    stop();                                    //
  }                                            //

  timeout = millis() + (rec_length * 1000);  // set stime after which the recording stops
}

// Arduino loop - copy data
void loop() {
  if (millis() < timeout) {
    copier.copy();                // record to wav file
  } else {                        //
    if (file) {                   // close file when done
      file.flush();               //
      Serial.print("File has ");  // output filesize in serial
      Serial.print(file.size());  //
      Serial.println(" bytes");   //
      file.close();               //
    }                             //
    stop();                       // stop the script from running when record is finished
  }
}
```
![rec_wav_hexdump](https://github.com/user-attachments/assets/78a471d3-19f7-49a5-8213-ed68f7d390c2)

**Comment by pschatzmann** [comment, marked as answer]
I never tried to record from an unconnected device, so I don't know what to expect.

You have a major bug in your logic: you cant just copy one format and pretend that it is another format!
Stick to one data format or perform a proper conversion as described in the Wiki...

**Reply by Tacco-172** [reply]
Thanks for your reply.

Please excuse me, of course it isn't going to work. I feel bad for not having been able to spot this for myself. I will test it by turning the quality of the ADC down and keeping it original until the save. Afterwards, I can (and probably will) still implement the converter. 

I didn't plug in any source into the ADC just to verify that the sketch itself does generally work. Therefore I expected the audio file to essentially be empty. The hissing noise however is very loud. I will from now on provide actual audio while experimenting.

**Comment by Tacco-172** [comment]
I tested it now and it works pretty well. Thanks for the thought-provoking impulse.

Now, importing the whole audiotools and audiodriver libraries take up a lot of space, 94% of flash in my case. I know audio processing is a rather hard exercise for an esp, but I still want to use it with other non-audio hardware. 
Is there a way to only include the sub-libraries needed for this sketch? I have an idea which ones to include, however: How can I make out the ones I need?
Or does another method of downsizing exist that I can use?

I plan to use a task-handler to pin the audioprocessing to a dedicated core to ensure 100% cpu time. The rest can easily be done on the other core.

**Comment by pschatzmann** [comment]
94 % ? I guess you just forgot to select the proper Partition Scheme! 
With Huge App it should be rather around 41%

**Reply by Tacco-172** [reply]
You're right. I had selected the D32 board in the IDE and therefore didn't have many partition schemes to choose from.
I switched over to the Dev Module, where every option is now visible, chose huge app and it dropped down to 39%. I selected this partition for now because you mentioned it and I dont have any experience with custom partition tables. Maybe in the future.
Huge thanks, now I can combine the hardware.
